client/maps/action-creators.ts

```diff
@@ -0,0 +1,221 @@
+import {
+  GetMapDetailsPayload,
+  GetMapsPayload,
+  MapInfoJson,
+  MapPreferences,
+  MapSortType,
+  MapVisibility,
+  Tileset,
+  UpdateMapPayload,
+  UpdateMapServerBody,
+} from '../../common/maps'
+import { ThunkAction } from '../dispatch-registry'
+import fetch from '../network/fetch'
+import { apiUrl } from '../network/urls'
+import { openSnackbar } from '../snackbars/action-creators'
+import { ClearMaps } from './actions'
+
+const upload = IS_ELECTRON ? require('./upload').default : null
+
+export function uploadLocalMap(path: string, onMapSelect: (map: MapInfoJson) => void): ThunkAction {
+  return async dispatch => {
+    if (!upload) {
+      return
+    }
+
+    dispatch({
+      type: '@maps/uploadLocalMapBegin',
+      payload: { path },
+    })
+
+    dispatch({
+      type: '@maps/uploadLocalMap',
+      payload: upload(path, apiUrl`maps`).then(({ map }: { map: MapInfoJson }) => {
+        if (onMapSelect) {
+          onMapSelect(map)
+        }
+
+        return map
+      }),
+      meta: { path },
+    })
+  }
+}
+
+export function getMapsList(
+  visibility: MapVisibility,
+  limit: number,
+  page: number,
+  sort: MapSortType,
+  numPlayers: number,
+  tileset: Tileset,
+  searchQuery: string,
+): ThunkAction {
+  return dispatch => {
+    const params = { visibility, limit, page, sort, numPlayers, tileset, searchQuery }
+
+    dispatch({
+      type: '@maps/getMapsBegin',
+      payload: params,
+    })
+
+    const reqUrl = apiUrl`maps?visibility=${visibility}&sort=${sort}&numPlayers=${JSON.stringify(
+      numPlayers,
+    )}&tileset=${JSON.stringify(tileset)}&q=${searchQuery}&limit=${limit}&page=${page}`
+
+    dispatch({
+      type: '@maps/getMaps',
+      payload: fetch<GetMapsPayload>(reqUrl),
+      meta: params,
+    })
+  }
+}
+
+export function toggleFavoriteMap(
+  map: MapInfoJson,
+  context: Record<string, unknown> = {},
+): ThunkAction {
+  return dispatch => {
+    const params = { map, context }
+
+    dispatch({
+      type: '@maps/toggleFavoriteMapBegin',
+      payload: params,
+    })
+
+    const reqUrl = apiUrl`maps/favorites/${map.id}`
+    dispatch({
+      type: '@maps/toggleFavoriteMap',
+      payload: fetch<void>(reqUrl, { method: map.isFavorited ? 'DELETE' : 'POST' }).then<void>(
+        () => {
```

Comment:
Should probably also be attached separately from the promise given to `dispatch` (see `regenMapImage` below)